### ISSUE TITLE
Request

### ISSUE THREAD

**Issue by d0ob** [InappropriatePenguin/combinator-toggle#1]
Hi there ! 

I'd like to make a request, or a help request : 

Some days ago I made a mod ( and what I mean by that is that I wanted a feature and someone nicely answered to help me make a mod ^^ )  for being able to switch a switch with a (ALT+)clic on it, without the need for going throught it's window. 

And today I see your mod ! I'd love to merge the two ! 

I removed the reach limit from yours and changed the keybind, so it works more like "mine" ; And now on the other hand, I'd love to add your sound effect to "mine" ! But being a full noob I don't know how to do that, my attempts are unsuccessful to say the least. ^^ 

I'd also like, since there is no reach limit for switches, and no more for combinators anymore, to make the sound audible from anywhere ingame, is it possible ? 

Could you help me doing all that ? 
In any case, thanks for your time, and thanks for your mod also ! 
Regards.

**Comment by InappropriatePenguin** [comment]
Thank you for your suggestion. I was actually planning to update the mod with the ability to toggle manual mode in trains and I've added support for power switches based on your request. Both of these are going to be optional and _off_ by default since they're, strictly speaking, outside the scope of the mod. I've already made those changes, but I need to test them.

The sound effect is easy enough to generate. You can call the `play_sound` function from the surface that the entity belongs to like so: `entity.surface.play_sound({position=entity.position, path=sound_path})`.

If you want the sound to be global (on that surface), all you need to do is _not_ pass a position argument. That's probably fine to do if you only care about single player, though I suspect it would be quite annoying in multiplayer.

The hotkey is easy enough for each player to customize to their liking, so I don't intend to change the default one. Not planning on removing the reach check for now, though it's simple enough to remove on your end (as you've done).

**Comment by d0ob** [comment]
Hi, thank you for your quick answer !

Yes I only play solo, for now atleast.
Indeed, I can imagine how it would be pretty annoying in multiplayer x)

Cool ! Well, if your update give the mod the same feature as 'mine", I'm not sure it's worth it to modify it ^^
I tried to copy that soundline from inside your code to inside "mine", but it prevent the game from lauching, saying me I screwed up somewhere ^^ I have to admit that ma coding level is zero, or lower. I read about modding the game when I wanted to make the feature, I was totally overwhelmed...

I'll be able to disable the reach limit (and that thanks to you saying we can in the mod's page, if not I don't think I would have figured out ^^) , and change the keybind (even more easly than I though, because I forgot the game automaticly gives a setting for that ^^) But you lost me for the sound distance !

**Comment by d0ob** [comment]
Saw the upload ! Nice, works like a charme ^^ Thanks for that ! Now my power swith is talkative when tickled =P

**Comment by InappropriatePenguin** [comment]
Glad you liked it :)

About the sound files, I was actually just referencing base game sound files, so my mod doesn't actually have any of its own. I'm looking for more appropriate sound effects for toggling power switches and locomotives though so that might change in the future.

Speaking of sound distance—since you only care about _you_ being to hear them from anywhere rather than the sounds being truly global for all players, there's a nicer way to do it that what I had previously suggested. Here's what you will need to change in the control.lua file:

Lines 19–22 should be changed from:
```lua
function play_sound_effect(entity, state, sounds)
    local sound_path = state and sounds["on"] or sounds["off"]
    entity.surface.play_sound({position=entity.position, path=sound_path})
end
```
to:
```lua
function play_sound_effect(player, state, sounds)
    local sound_path = state and sounds["on"] or sounds["off"]
    player.play_sound({path=sound_path})
end
```

And then in lines 46, 49, and 53, replace the word `entity` with `player`. So the following:
```lua
play_sound_effect(entity, control.enabled, sound_paths)
```
should be changed to:
```lua
play_sound_effect(player, control.enabled, sound_paths)
```
in each of those lines. Let me know if you run into any issues making those changes. If not, you can go ahead and close the issue.

**Comment by d0ob** [comment]
It works ! Damn it works x) It's so sweet ! Thanks much ! 

Since some days I have fun translating mods, and it's a kindof "thank you" I'm able to do for modders ^^
There's not much to do here but/so I couldn't not do it ^^ You'll see if you do anything with it. 

_I use the english Factorio but I also installed a french one to look at legit words, it's very talkative compared to english so here I prefered translating the options in description than in their name itself. I tested ingame and it works nicely !_ 

I also finally learned how to use pull request on github so I'll even do it by the legit way !

**Comment by InappropriatePenguin** [comment]
Good! I'm glad it worked.

And thank you for taking the time to do that translation :) I've already merged your pull request and it will be in the next release. That probably won't be for a few more days since I just pushed one out today. Maybe by then I will have found some custom sound effects that I like.

I'm going to go ahead and close the issue then. Do feel free to start a new one if there's anything else.
